### ISSUE TITLE
Intersect() truncation fix and loss of precision

### ISSUE THREAD

**Issue by erikh2000** [Turfjs/turf#1213]
I'm familiar with the code inside intersect(). And I see a clever hack contributed by @DenisCarriere in this commit: https://github.com/Turfjs/turf/commit/66722db5bc5c461b3293183e75f6f781be7a435a#diff-15768e5e1d39d953dd9dfb38dc6ff146

I'm sure it's not news to you, that this can cause some real problems. Basically, I've got an app that needs 6-digit precision, and the hack truncates to 4 digits. Comically, the end result of this truncation for me would be automated farm equipment planting seeds in fields with giant gaps of unplanted areas that were created by those two lines of code! I've attached an example pic of the difference between the original polygon for a planting zone in a field, overlaid with the simplified version used by the hack in intersect().

I will congratulate Denis in that the hack *does* get around the TopologyException from JSTS that has been plaguing us for years. But since I can't use that low level of precision, and I also need the TopologyException problem to be fixed, it leaves me in a place where I'm considering dropping use of Turf intersect(). (not Turf entirely--just functions like intersect() that don't seem like they could possibly work for what I need to do)

Please correct my understanding above if I've got something wrong.

I know there is a planned effort to replace JSTS, and y'all want to be out of the business of fixing/working-around JSTS. Makes perfect sense. 

If practical, I'd like to help with the part of this that replaces intersect(). Or if there is some code that is already written, but maybe has some caveats/unhandled edge cases, I don't mind looking at that and trying to improve. I've got some time this month. I have the build environment for Turf working and understand how to contribute changes.

![image](https://user-images.githubusercontent.com/8631070/34628339-bfd901f8-f218-11e7-9293-008c1d0dd2d1.png)

**Comment by DenisCarriere** [comment]
@erikh2000 Thanks for reporting this issue, maybe an easy solution is by adding `validate` as an optional parameter.

I believe we've already talked about this in the GitHub issues, a way for the user to prevent any validation on the data input, that way it would speed things up if you don't need to validate the input data.

@erikh2000 Are you comfortable sending a PR with the change? I can publish it as a patch for `@turf/intersect`.

```js
function intersect(poly1, poly2, options) {
    options = options || {};
    
    var geom1 = getGeom(poly1);
    var geom2 = getGeom(poly2);

    // Return null if geometry is too narrow in coordinate precision
    // fixes topology errors with JSTS
    // https://github.com/Turfjs/turf/issues/463
    // https://github.com/Turfjs/turf/pull/1004
    if (options.validate === true) {
        if (cleanCoords(truncate(geom2, {precision: 4})).coordinates[0].length < 4) return null;
        if (cleanCoords(truncate(geom1, {precision: 4})).coordinates[0].length < 4) return null;
    }
...
```

**Comment by erikh2000** [comment]
@DenisCarriere, thanks for the reply.

I think the .validate option would help some people. But in my case, I'll just be back to having topology exceptions if I don't run cleanCoords(truncate()) on the coordinates. The starting polygon I have above is valid (correct winding, no kinks, no duplicate coords) and was generated from an earlier call to difference(). It's getting tripped up inside of JSTS intersection() around that sliver of a peninsula on the left side of the shape. These types of problems will come up on a regular basis as we slice and dice literally millions of polygons--no debug error in the app required.

I can work on a replacement for @turf/intersect that doesn't use JSTS. I was thinking of adapting from Luis Lafuente's project https://github.com/llafuente/js-2dmath/blob/master/lib/intersection.js (MIT license). Another possibility is working with @morganherlocker 's (abandoned?) Weiler-Atherton project https://github.com/morganherlocker/weiler-atherton . It would be nice to start with some code that is already known to pass all the edge cases one tends to overlook. Of course, we can beef up the unit tests in intersect() to make sure.

Anyhow, advice for a starting point is appreciated, particularly if Morgan's work seems usable. Or if y'all are not quite ready for this kind of replacement, that's fine too. I can work on an off-Turf solution for myself.

**Comment by rowanwins** [comment]
Hi @erikh2000 

Currently the most promising option for removing jsts is [martinez](https://github.com/w8r/martinez) which we've been trying to progress with its author. It's certainly getting pretty close, although we've still got one or two edge cases that we haven't yet cracked (cases that would likely break jsts as well).

I'm a bit tied up for the next week but after that I plan on seeing if I can replace jsts in union, intersect, difference in turf.

**Comment by erikh2000** [comment]
@rowanwins Nice. Thanks for the update. I may play with that code a bit.

**Comment by rowanwins** [comment]
Likely resolved since v6.2.0 which no longer truncates coords
